Implement stateless AAD authentication filter which uses appRoles (#494)

wmitzel-airplus opened the pull request:
## Summary
This pull request implements a stateless authentication filter that uses `id_token` and the `appRole` feature of AAD. Some details and motivation why this makes sense have been discussed in #494. 

## Issue Type
 - New Feature

## Starter Names
  - active directory spring boot starter

## Additional Information
I've created a new authentication filter which is similar to the already existing `AADAuthenticationFilter`. However, it does not use `HttpSession` and thus can be used for stateless Spring backends. The basic JWT claims are checked by the `UserPrincipalManager`. Additionally, the `roles` scope is read and the roles are converted to `GrantedAuthory` which are then attached to the principal.

I've also added an example project which demonstrates the usage.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/azure-spring-boot?pullRequest=512) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/azure-spring-boot?pullRequest=512) before we can accept your contribution.<br/><br/>:x: wmitzel-airplus [sign now](https://cla.opensource.microsoft.com/Microsoft/azure-spring-boot?pullRequest=512)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/azure-spring-boot?pullRequest=512) it.</sub>

wmitzel-airplus commented:
Hi @Incarnation-p-lee 
I tried to follow [HowToContribute.md](https://github.com/Microsoft/azure-spring-boot/blob/master/HowToContribute.md) where it mentions a passing build, updated documents, sample code, readme and code coverage. Maybe it makes sense to add the recommendation regarding scope and size of a PR to this guideline.

I agree that working on smaller PR can lead to more efficiency. This PR currently consists of 3 parts:
1) the actual implementation in `azure-spring-boot/` (~159 lines containing the actual change plus ~155 lines of tests). I would argue that splitting the implementation further down makes only little sense as it only consists of one Filter that parses the token to get the roles.
2) update of documentation in `azure-spring-boot-starters` (30 lines)
3) an example project demonstrating the new feature in `azure-spring-boot-samples/azure-active-directory-spring-boot-app-role-sample` (~290 lines + 80 lines readme)

I would suggest to split it into two PRs: 
* the first PR (this one) containing 1) and 2): implementation + basic documentation
* the second PR holding the example project 3). I would extract the example project from this PR and will create a new one for it.

Incarnation-p-lee commented:
@wmitzel-airplus 
Great, following the rule do one thing at one time, I suggest you can separate into 3 PRs.
* Implementation code including unit test
* Sample code.
* Documentation for sample code, like README... etc.
Just personal suggestion, you can make you own choose there.

wmitzel-airplus commented:
Hi, sorry for the long delay. Today is the first day of work after my Christmas leave. I've removed the sample module and will add it as a separate PR later.
I am wondering whether I should rebase my changes onto master before we go on?

Incarnation-p-lee commented:
@wmitzel-airplus 
Thanks for your effort and contribution. You can just merge master tree code when your PR is ready. Please let me know if your PR is ready. Thanks again.

wmitzel-airplus commented:
@Incarnation-p-lee 
I've updated the PR and it's now ready for review.

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationProperties.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.validation.annotation.Validated;
+
+@Validated
+@ConfigurationProperties("azure.aad.app-role")
```
please follow the naming conversion as AADAuthenticationProperties, use `azure.activedirectory` instead of `azure.aad.app-role`.

wmitzel-airplus replied:
The property "namespace" `azure.activedirectory` is already bound to `com.microsoft.azure.spring.autoconfigure.aad.AADAuthenticationProperties`. I did not manage to use `azure.activedirectory.app-role` for my new configuration without having side effects of `AADAuthenticationProperties` being instantiated and beans being created automatically. As I didn't want to break the existing feature I've chosen this different property name. 
Is there a way of doing it without the side effects I've described?

wmitzel-airplus commented on azure-spring-boot-starters/azure-active-directory-spring-boot-starter/README.md:
```diff
@@ -91,6 +91,36 @@ private AADAuthenticationFilter aadAuthFilter;
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        [...]
+        http.addFilterBefore(appRoleAuthFilter, UsernamePasswordAuthenticationFilter.class);
```
Yes, you are right. For this filter to work in a stateless way, we need to set the session creation policy as you have described.

wmitzel-airplus commented on azure-spring-boot-starters/azure-active-directory-spring-boot-starter/README.md:
````diff
@@ -91,6 +91,36 @@ private AADAuthenticationFilter aadAuthFilter;
+    }
+```
+
+* Role-based Authorization with annotation `@PreAuthorize("hasRole('rolename')")`
````
Is this a valid issue? I've used the same formatting as in the rest of the README.md files. They usually use 1 space indent.

wmitzel-airplus commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,86 @@
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.proc.BadJOSEException;
+import com.nimbusds.jwt.proc.BadJWTException;
+import com.nimbusds.oauth2.sdk.util.CollectionUtils;
```
Will fix.

wmitzel-airplus commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,86 @@
+        if (authHeader != null && authHeader.startsWith(TOKEN_TYPE)) {
+            try {
+                final String token = authHeader.replace(TOKEN_TYPE, "");
+
```
This was just a personal style decision. I can remove this.

wmitzel-airplus commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,86 @@
+            }
+        }
+
+        filterChain.doFilter(request, response);
```
What type of "cleanup" do you have in mind? 
In the default configuration, Spring uses a `ThreadLocal` to store the security context and also makes sure to clear it once the request is processed.

Incarnation-p-lee replied:
I mean set the security context to null after request processed to make sure the stateless. Yes the spring security framewrok will did that for us if we config stateless session policy. But from our filter review, we cannot trust the user config and need to make sure the context is stateless. How do you think of it ?

wmitzel-airplus replied:
I was referring to the first paragraph of [SecurityContextHolder, SecurityContext and Authentication Objects](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#securitycontextholder-securitycontext-and-authentication-objects) in the spring security docs where it says:
> [...] Using a ThreadLocal in this way is quite safe if care is taken to clear the thread after the present principal’s request is processed. Of course, Spring Security takes care of this for you automatically, so there is no need to worry about it.

I've validated this behaviour and seen that this code cleans up each request: https://github.com/spring-projects/spring-security/blob/f234a5fbdbf43dbceecfdfe3d41f10b09aa5f197/web/src/main/java/org/springframework/security/web/context/SecurityContextPersistenceFilter.java#L111-L116

But as a defense in depth measure, we can clear the security context after the `doFilter` call. I suggest to do:
```
        boolean cleanupRequired = false;
        [...]
        SecurityContextHolder.getContext().setAuthentication(authentication);
        cleanupRequired = true;
        [...]
        try {
            filterChain.doFilter(request, response);
        } finally {
            if (cleanupRequired) {
                //Clear context after execution
                SecurityContextHolder.clearContext();
            }
        }
```
What do you think?

Incarnation-p-lee replied:
Yes, Spring security will do that for you as you mentioned in manual but I mean the authentication in spring context. Sorry for misleading here.

It is OK that if user configuration sessionless, here I just want to make sure the authentication of context is clear before the session filter process(only if user forget to enable stateless policy). Like this `SecurityContextHolder.getContext().setAuthentication(null);`

Even if use forget to enable stateless, the `SecurityContextHolder.getContext().setAuthentication(null);` will ensure the null authentication will be store to repo in below code.

```java
		finally {
			SecurityContext contextAfterChainExecution = SecurityContextHolder
					.getContext();
			// Crucial removal of SecurityContextHolder contents - do this before anything
			// else.
			SecurityContextHolder.clearContext();
			repo.saveContext(contextAfterChainExecution, holder.getRequest(),
					holder.getResponse()); // clean auth before finally to make sure auth in context is null
			request.removeAttribute(FILTER_APPLIED);

			if (debug) {
				logger.debug("SecurityContextHolder now cleared, as request processing completed");
			}
		}
```

wmitzel-airplus commented:
> @Incarnation-p-lee
> As I understand, if we leverage stateless filter we cannot use normal session based auth together. So from AuthProperty you may need to add some option like statelessEnabled to tell this. If then you can add configuration to AuthProperty instead of inherit from it.

Indeed I've not foreseen using this filter together with the current session based auth. Would this be a typical scenario?

I tried not to break or influence any existing feature with this new authentication method.  One place where I did not wholly succeed with this intention is `AADAppRoleAuthenticationProperties` extending `AADAuthenticationProperties`. Here I felt that there is a lot of default configuration that I didn't want to copy.

If I understand you correctly, the plan is to remove `AADAppRoleAuthenticationProperties`, add a property `statelessEnabled` (or maybe `useAppRoles` to reflect the class name?) to `AADAuthenticationProperties` and either create `AADAuthenticationFilter` or `AADAppRoleAuthenticationFilter` in the auto configuration?

wmitzel-airplus commented on azure-spring-boot-starters/azure-active-directory-spring-boot-starter/README.md:
````diff
@@ -91,6 +91,39 @@ private AADAuthenticationFilter aadAuthFilter;
+    }
+```
+
+* Role-based Authorization with annotation `@PreAuthorize("hasRole('rolename')")`
````
Is this a valid issue? The other list items are also indented with 1 space.

Incarnation-p-lee replied:
The codacy-robot did this automatically, I may adjust the rule for this if most leverage 1 space.

Incarnation-p-lee replied:
I should remove this rule from codacy.

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,86 @@
+                SecurityContextHolder.getContext().setAuthentication(authentication);
+            } catch (BadJWTException ex) {
+                log.warn("Invalid JWT: " + ex.getMessage());
+                throw new ServletException(ex);
```
For log message for exception, my suggestion may looks like this `throw new Exception("Invalid JWT: Expired xxxxx", ex);` which include some message to user, as well as the cause.

wmitzel-airplus replied:
I am not sure I understand what you mean. When this exception is catched, `ex.getMessage()` will contain either `Expired JWT` or `JWT before use time`. From my point of view, there is no additional information I can provide in order to make more sense out of this exception. Or do you mean the fact that here `throw new ServletException(ex);`  no message is specified and only available in the causedBy?

Incarnation-p-lee replied:
I mean if there are some message here, we can get the hint from log instead of debugging or something like that `ex.getMessage()` to find the cause. It is not big issue here, just from the users view about the exception throw.

Incarnation-p-lee commented:
@wmitzel-airplus 
Yes, I think it may not possible to leverage both session and session stateless filter in a spring web application, as they should be exclusive to each other.

For the properties, I think both session and session stateless are in the scope of AADProperties. By default we may leverage the session filter, if stateless enabled we should create other filters to handle the stateless scenario.

In current implementation of aad, there are two autoconfiguration because one is for front-end auth and the other is for backend auth.

Here is some sample for session stateless property 
https://github.com/Microsoft/azure-spring-boot/blob/2c6c1470262367006de1d4db831f67b4c94528b5/azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CProperties.java#L84

As well as the related auto-configuaration
https://github.com/Microsoft/azure-spring-boot/blob/2c6c1470262367006de1d4db831f67b4c94528b5/azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CAutoConfiguration.java#L128

And the key word for this scenario should be stateless, and the `azure.activedirectory.app-role` may have no hint to user that we are leveraging stateless session.

Thanks all for you great work. Feel free to contact us if there  is something incorrect I stated in below.

wmitzel-airplus commented:
Okay, will do the property and auto configuration change soon and update the PR.

wmitzel-airplus commented:
I've updated the PR and combined the auto-configuration classes. There is now `azure.activedirectory.session-stateless` which is `false` by default and if `true` activates the new feature.
Furthermore, I added a cleanup of the `SecurityContext` after the filtering has been done and adapted the tests to check this cleanup.

strivedi-mrc commented:
Hi, I have run into exact same issue on a current integration project using Azure AD as the auth provider to support client credentials Oauth2.0 flow. Can you please let me know when this PR can be merged to master.  Thanks wmitzel-airplus for addressing this issue in your PR.

wmitzel-airplus commented:
I've updated the PR with the following changes:
* Explicit audience check for the stateless filter. While there is an implicit check of the audience in the existing filter by calling the graph API, this check was missing for the stateless filter.
* Adapt the user group check so that it gives a warning when user groups are defined in a stateful scenario.

Additionally, I've created a small demo app which demonstrates the usage of this feature. You can find it here: wmitzel-airplus/azure-active-directory-spring-boot-app-role-sample@cd0f502

Incarnation-p-lee commented:
@wmitzel-airplus
Thanks a lot for your contribution. Could you please separate this big PR into 2 or 3 ? Like feature PR, configuration PR, and test PR. As production project, it is not easy to merge such big PR unless you have strange reason to insist that.

And for sample, you can add one `stateless` sample to our repo sample module.
Again, very appreciate for your contribution.

wmitzel-airplus commented:
> @wmitzel-airplus
> Thanks a lot for your contribution. Could you please separate this big PR into 2 or 3 ? Like feature PR, configuration PR, and test PR. As production project, it is not easy to merge such big PR unless you have strange reason to insist that.
> 
> And for sample, you can add one `stateless` sample to our repo sample module.
> Again, very appreciate for your contribution.

I checked your proposal, and I'm not sure it would help. The configuration is only 2 classes and together 55 lines. Most of the PR are tests (312 lines) for the feature. 

If it simplifies the review, I split the implementation and the tests into two PRs. Personally, I would not accept a PR without tests. But this is up to you to decide.

Incarnation-p-lee commented:
Thanks for your effort. For long term maintains we suggest to do one thing for one PR. I think you can separate the implementation code and test code in 2 PRs. Your effort for this `stateless` feature indicates that you have great sense of responsibility of your code.

Thanks again for your contribution, very appreciate.

lhanson commented:
From the peanut gallery, it seems odd to separate tests into their own PR since to my mind they're intrinsically linked to the "one thing" the code is actually doing. Additionally, for long term maintenance it seems like it would complicate things like performing a `git bisect` to retroactively find issues if the tests are in a separate commit from the code they're exercising.

In any case, thanks to everyone for their work on this, I'm eager to see it in master.

wmitzel-airplus commented:
Thanks for your patience. I've extracted the tests to PR #684.

patfrat commented:
PreAuthorize hasRole with roles is logical  
The stateless is good too  
But i have no more groups attached to my user and i need to verify if the user is member of group with isMemberOf() method

wlami commented:
> But i have no more groups attached to my user and i need to verify if the user is member of group with isMemberOf() method

This is intended - as with big organizations you get very many groups at some point you cannot fit them all into your JWT. Thus the appRole feature allows you to define roles which are specific to *your* application. What you can do then is assigning the appRole to an AAD group in "Enterprise Application --> Users and Groups". You can check the example at https://github.com/wmitzel-airplus/azure-active-directory-spring-boot-app-role-sample .
I see no easy way of having the statelessness and group claims at the same time. If you need group claims you can use the existing AAD filter.

patfrat commented:
> I see no easy way of having the statelessness and group claims at the same time. If you need group claims you can use the existing AAD filter.

I have done a fork to suite my needs with a new branch and this commit :

https://github.com/patfrat/azure-spring-boot/commit/67791818c9921fc7fa05482ee9dd4564f49428ce

And it works ! No more JSESSIONID ! Stateless ;)

wlami commented:
> > I see no easy way of having the statelessness and group claims at the same time. If you need group claims you can use the existing AAD filter.
> 
> I have done a fork to suite my needs with a new branch and this commit :
> 
> [patfrat@6779181](https://github.com/patfrat/azure-spring-boot/commit/67791818c9921fc7fa05482ee9dd4564f49428ce)
> 
> And it works ! No more JSESSIONID ! Stateless ;)


Yes, stateless, but did you measure your latencies after changing the filter? In https://github.com/patfrat/azure-spring-boot/commit/67791818c9921fc7fa05482ee9dd4564f49428ce#diff-c5303eefb9fd562944ee941237e01c28R70 and in https://github.com/patfrat/azure-spring-boot/commit/67791818c9921fc7fa05482ee9dd4564f49428ce#diff-c5303eefb9fd562944ee941237e01c28R73 you are executing 2 remote REST calls for each incoming request you want to authenticate.

I checked the same approach but it was not an option for me because of the added latency. From my understanding, the reason why the stateful filter is stateful is the fact that it caches the two remote calls so that they are only required once for the session.

The alternative solution in this PR does not require any additional 3rd party calls for authentication and validation of the token which makes it stateless and performant.

patfrat commented:
> it caches the two remote calls so that they are only required once for the session.

I understand the need for performance. With the stateful approach, my problem is that after a logout from the angular frontend, and a new login with another user, the previous session is still active ! Even when I send a logout request to the backend that is supposed to destroy the session, I stay on the previous user until I exit the browser.

I have missed something ?

wlami commented:
> I have missed something ?

Yes, with the stateful filter it is not sufficient to send another JWT with a different JWT inside. It will not reevaluate the JWT as long as there is a JSESSIONID which your browser remembers until you don't close it. You need to call the `/logout` endpoint which spring security adds per default (or your custom logout endpoint) so that your server-side JSESSIONID gets invalidated.

patfrat commented:
> You need to call the `/logout` endpoint which spring security adds per default (or your custom logout endpoint) so that your server-side JSESSIONID gets invalidated.

It does not seems enough. I will try another way by comparing the JWT received !

patfrat commented:
This approach works for me while keeping the session for performance purposes

https://github.com/patfrat/azure-spring-boot/commit/b8fbc8f837602642770888869d1a68f35e1fd87d

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,91 @@
+
+    private static final Logger log = LoggerFactory.getLogger(AADAppRoleAuthenticationFilter.class);
+
+    private static final String TOKEN_HEADER = "Authorization";
```
Please use `HttpHeaders.AUTHORIZATION`.

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,91 @@
+    private static final JSONArray DEFAULT_ROLE_CLAIM = new JSONArray().appendElement("USER");
+    private static final String ROLE_PREFIX = "ROLE_";
+
+    private UserPrincipalManager principalManager;
```
final ?

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java:
```diff
@@ -0,0 +1,91 @@
+        final String authHeader = request.getHeader(TOKEN_HEADER);
+        boolean cleanupRequired = false;
+
+        if (authHeader != null && authHeader.startsWith(TOKEN_TYPE)) {
```
hasText instead of != null ?

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilterAutoConfiguration.java:
```diff
@@ -5,37 +5,38 @@
 @Configuration
 @ConditionalOnWebApplication
-@ConditionalOnProperty(prefix = "azure.activedirectory", value = {"client-id", "client-secret"})
+@ConditionalOnProperty(prefix = "azure.activedirectory", value = {"client-id"})
```
leverage below `PROPERTY_PREFIX` ?

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java:
```diff
@@ -49,7 +57,16 @@ public UserPrincipalManager(JWKSource<SecurityContext> keySource) {
+        this.aadAuthProps = aadAuthProps;
+        this.explicitAudienceCheck =  explicitAudienceCheck;
+        if (explicitAudienceCheck) {
+            //client-id for "normal" check
```
space after '//'

Incarnation-p-lee commented on azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java:
```diff
@@ -49,7 +57,16 @@ public UserPrincipalManager(JWKSource<SecurityContext> keySource) {
+                                ResourceRetriever resourceRetriever,
+                                boolean explicitAudienceCheck) {
+        this.aadAuthProps = aadAuthProps;
+        this.explicitAudienceCheck =  explicitAudienceCheck;
```
duplicated space after `=`

wmitzel-airplus commented:
Thank you very much for the code review. I've fixed all of the issues.

Incarnation-p-lee commented:
@wmitzel-airplus
Thanks a lot for your great work and contribution, very appreciate. I think the code is ready to merge and only one thing about the filter name. As it is session stateless, I think the filter should emphasize this, like `AADRoleAuthenticationStatelessFilter.java`,how do you think of it ?

wmitzel-airplus commented:
> @wmitzel-airplus
> Thanks a lot for your great work and contribution, very appreciate. I think the code is ready to merge and only one thing about the filter name. As it is session stateless, I think the filter should emphasize this, like `AADRoleAuthenticationStatelessFilter.java`,how do you think of it ?

I like the idea of adding "Stateless" to the name of the filter. However, I'd prefer not to abbreviate "AppRole" to "Role" as the official feature in AAD is called "approle". If you search on the internet for the term "app role" you directly get to the documentation on how to add application roles to the manifest of your application registration.
So my suggestion would be `AadAppRoleStatelessAuthenticationFilter`. What do you think?

Incarnation-p-lee commented:
sounds good, please help to update the name.

Incarnation-p-lee commented:
Thanks a lot but could u please following the convention use `AAD` instead of `Aad` ? as well as update the name of REAMD.ME ?

Incarnation-p-lee commented:
Thanks a lot for you contribution, very appreciate. Could you please help to add sample code for this scenario ? like [this](https://github.com/microsoft/azure-spring-boot/tree/d1546863ad8a5f7950c0cce881a124d7bf6d398c/azure-spring-boot-samples/azure-active-directory-b2c-oidc-spring-boot-sample)

geoaxis commented:
👍  
PS: when can we expect a maven release for this?

strivedi-mrc commented:
Any idea on the ETA to make a maven release for this code in master ?
